QQE-1584.md

```diff
@@ -0,0 +1,82 @@
+# QQE-1584 Quarkus LangChain4J testing
+
+JIRA link: https://issues.redhat.com/browse/QQE-1584
+
+This is a high-level strategy for testing of langchain4j
+
+## Scope of the testing
+@SystemMessage + @UserMessage annotations
+RAG functionality
+Integration with MCP API
+
+### Optional, but desirable:
+@Tool + @RegisterAiService (only supported for OpenAI, so cannot be tested locally)
+Observability for tool requests (to avoid a situation, where LLM overwhelms the database with too many requests)
+
+## Existing coverage upstream:
+### Workshop
+https://quarkus.io/quarkus-workshop-langchain4j/step-11/#running-the-llm-inference-locally  and  https://github.com/quarkusio/quarkus-workshop-langchain4j/tree/main/step-11
+(they are not run automatically, as far as I can tell)
+The tests use jlama, which requires Java 21+
+Attempt to run the test locally with jlama fails on my laptop(32.0 GiB RAM, 13th Gen Intel® Core™ i7-1365U × 12). Will try to run it on server, but I am not very optimistic about that.
+Successfully ran it on the server. The downside is there is no rest api, and we have to connect through websocket (which is not optimal). Also, it hangs regularly, even on xlarge.
+
+### Quakriverse extension:
+There are integration tests modules in the extension, but they contain no tests at the moment.
+According to Mario, this is intentional, since they consider upstream lang4chain tests to be enough (but we do not!)
+Ollama modules can be started on the laptop in dev mode. Jlama ones can be run as well.
+
+## Automated test development
+### Functional testing
+1. Add some tests for existing functionality into https://github.com/quarkiverse/quarkus-langchain4j/tree/main/integration-tests/ollama, https://github.com/quarkiverse/quarkus-langchain4j/tree/main/integration-tests/simple-ollama and https://github.com/quarkiverse/quarkus-langchain4j/tree/main/integration-tests/jlama
+2. Beware, that these modules use models of different size, and this can affect reproducibility of tests. Some tests may not work on our hardware with some of the models.
+3. Investigate, whether it is possible to have RAG with ollama/jlama (it should be: [1][2]), add some new functionality and coverage for it into integration tests (maybe in new module)
+4. Add some observability tests.
+5. Add ability to change version of Quarkus and quarkus-langchain4j, which is used in integration tests. Adding new property for version in module pom should be enough.
+6. MR should target main, the reviewer(s) should be Mario and/or Georgios, since they both agreed, that this is the way to go.
+7. Try running the tests on our test servers. If it fails, we may want to try running ollama on a separate machine and connect to it. 
+8. Create a new Jenkins job for prod testing on x86_64 in "general" scope and one for ARM_64 in "extended" scope.
```

Comment:
Does this mean we're going to run the extensions' test suites with productized Quarkus? Is this possible? Or am I misreading this?

Reply:
Again, maybe https://github.com/quarkus-qe/quarkus-test-plans/pull/237/files#r2180717349 is related.

But where would this change happen? Is this something that needs to be merged upstream to be done as we'll need this for support? This is not clear at all from the document either.